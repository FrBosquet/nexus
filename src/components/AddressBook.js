import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import ContactList from './ContactList';
import SearchBar from './ContactList/SearchBar';
import Layout from './Common/Layout';
import StatusBar from './Common/StatusBar';

import Contacts from '../services/contacts';

class AddressBook extends Layout {
  static propTypes = {
    ...Layout.propTypes,
    className: PropTypes.string,
  };

  state = { contacts: [], filter: '' };

  async componentDidMount() {
    this.setState({ contacts: await Contacts.read() });
  }

  getFilteredList() {
    const { contacts, filter } = this.state;
    const regexp = new RegExp(filter, 'i');

    return contacts.filter(({ name: { first, last } }) =>
      regexp.test(`${first} ${last}`),
    );
  }

  updateFilter = e => this.setState({ filter: e.target.value });

  render() {
    const { className } = this.props;
    const { filter } = this.state;
    const contacts = this.getFilteredList();
    const element = super.render();

    if (!element) {
      return null;
    }

    return (
      <main className={className}>
        <StatusBar />
        <SearchBar value={filter} onChange={this.updateFilter} />
        <ContactList items={contacts} />
        {element}
      </main>
    );
  }
}

export default styled(AddressBook)`
  padding-top: 5rem;
`;
