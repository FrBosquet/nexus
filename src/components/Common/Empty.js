import React from 'react';
import { rgba } from 'polished';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { fromTheme } from '../../utils/styled';

const Empty = ({ className }) => <div className={className} />;

Empty.defaultProps = {
  className: '',
};

Empty.propTypes = {
  className: PropTypes.string,
};

export default styled(Empty)`
  @media (${fromTheme('--screen-medium')}) {
    background: ${fromTheme('--color-light')};
    border-left: 1px solid ${props => rgba(props.theme['--color-dark'], 0.1)};
    height: 100vh;
    left: 32rem;
    position: fixed;
    top: 0;
    width: calc(100% - 32rem);
  }
`;
