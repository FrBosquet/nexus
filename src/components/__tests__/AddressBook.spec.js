import React from 'react';
import { Route } from 'react-router-dom';
import {
  render,
  cleanup,
  waitForElement,
  fireEvent,
} from 'react-testing-library';
import Empty from '../Common/Empty';

import AddressBook from '../AddressBook';
import withProvider from '../../utils/hocs';

import Contacts from '../../services/contacts';

jest.mock('../../services/contacts');
global.window.scrollTo = () => null;

describe('Address Book', () => {
  const Enhanced = withProvider(() => (
    <AddressBook>
      <Route component={() => <Empty />} />
    </AddressBook>
  ));
  const contacts = [
    {
      picture: { thumbnail: '' },
      name: { first: 'Adrian', last: 'de la Rosa' },
      id: '0',
    },
    {
      picture: { thumbnail: '' },
      name: { first: 'Marta', last: 'Fonda' },
      id: '1',
    },
    {
      picture: { thumbnail: '' },
      name: { first: 'Mikel', last: 'Rumayor' },
      id: '2',
    },
  ];

  beforeEach(() => {
    Contacts.read.mockReturnValue(Promise.resolve(contacts));
  });

  afterEach(() => {
    cleanup();
  });

  it('renders by default', () => {
    const { container } = render(<Enhanced />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('calls Contact.read on mount', () => {
    render(<Enhanced />);
    expect(Contacts.read).toHaveBeenCalled();
  });

  it('renders a card for each contact', async () => {
    const { getByText } = render(<Enhanced />);
    await waitForElement(() => getByText('Marta Fonda'));
    expect(getByText('Adrian de la Rosa')).toBeDefined();
    expect(getByText('Marta Fonda')).toBeDefined();
    expect(getByText('Mikel Rumayor')).toBeDefined();
  });

  it('renders a header for each initial letter', async () => {
    const { getByText } = render(<Enhanced />);
    await waitForElement(() => getByText('Marta Fonda'));
    expect(getByText('a')).toBeDefined();
    expect(getByText('m')).toBeDefined();
  });

  it('scrolls to a heading when clicking on it', async () => {
    const { getByText } = render(<Enhanced />);
    await waitForElement(() => getByText('Marta Fonda'));
    const spy = jest.spyOn(global.window, 'scrollTo');
    getByText('a').click();
    expect(spy).toHaveBeenCalled();
  });

  it('filters the list when typing on the search bar', async () => {
    const { getByTestId, getByText } = render(<Enhanced />);
    await waitForElement(() => getByText('Marta Fonda'));
    const input = getByTestId('input');
    fireEvent.change(input, { target: { value: 'Marta' } });
    expect(() => getByText('Mikel Rumayor')).toThrow();
    expect(() => getByText('Adrian de la Rosa')).toThrow();
  });
});
