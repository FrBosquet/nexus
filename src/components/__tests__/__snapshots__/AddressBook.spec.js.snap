// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Address Book renders by default 1`] = `
<main
  class="sc-jzJRlG kAhIWh"
>
  <section
    class="sc-gqjmRU jPNVfR"
  >
    <input
      data-testid="input"
      type="text"
      value=""
    />
    <i
      class="sc-gZMcBi giNnHm material-icons"
    >
      search
    </i>
  </section>
  <ul
    class="sc-iwsKbI jHhKfw sc-dnqmqq bDsesi"
  />
  <div
    class="sc-bdVaJa etPSam"
  />
</main>
`;
